### PR TITLE
Sponsored members implemented

### PR DESCRIPTION
**Pull request by martin-kuba**
- added a boolean flag "sponsored" to class Member and table MEMBERS
- added a table MEMBERS_SPONSORED for link between sponsored member and sponsoring user
- added 5 API calls to MembersManagerMethod: createSponsoredMember, sponsorMember, extendExpirationForSponsoredMember, getSponsoredMembers, getSponsors
- added corresponding 4 methods to MembersManager/MembersManagerEntry and 1 to UsersManager/UsersManagerEntry (getSponsors)
- added corresponding 4 methods to MembersManagerBl/MembersManagerBlImp and 1 to UsersManagerBl/UsersManagerBlImpl
- added 5 methods for managing SPONSOR role to VosManagerBl: isUserInRoleForVo, addSponsorRole (for user and group) and removeSponsorRole (for user and group)
- implemented the same 5 methods in AuthzResolverImpl and rewritten other methods for manipulating user or group roles for VOs to use generic addVoRole and removeVoRole as well
- expiration rules are implemented in MembersManagerBlImp.removeSponsor() and VosManagerBlImpl.removeSponsorRole()

### CODE REVIEW

**Comment by martin-kuba** [comment]
- added API calls VosManagerMethod for addSponsorRole and removeSponsorRole
- tested calling the two calls and API call createSponsoredMember, CLI clients are available at https://github.com/martin-kuba/perun-java-cli

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java [hunk 1]
```diff
@@ -219,34 +206,31 @@ public int compareByModifiedAt(Auditable auditable) {
...
 	@Override
 	public String toString() {
-		StringBuilder str = new StringBuilder();
-
-		return str.append(this.getClass().getSimpleName()).append(":[").append(
-				"id='").append(this.getId()).append('\'').append(
-				']').toString();
+		return this.getClass().getSimpleName() + ":[" +
+				"id='" + this.getId() + '\'' +
+				']';
```
Comment (licehammer) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java]: What is the reason for this? String builder should be more efficient, shouldn't it?

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java [hunk 2]
```diff
@@ -40,7 +37,7 @@ public User(int id, String firstName, String lastName, String middleName, String
 	}
 
 	public User(int id, String firstName, String lastName, String middleName, String titleBefore, String titleAfter,
-			String createdAt, String createdBy, String modifiedAt, String modifiedBy, Integer createdByUid, Integer modifiedByUid) {
+	            String createdAt, String createdBy, String modifiedAt, String modifiedBy, Integer createdByUid, Integer modifiedByUid) {
```
Comment (licehammer) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java]: Please, use \t for indentation.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java [hunk 3]
```diff
@@ -2099,6 +2085,120 @@ public void sendPasswordResetLinkEmail(PerunSession sess, Member member, String
...
+
+	private void setLoginToSponsoredUser(PerunSession sess, User sponsoredUser, String loginAttributeName, String login) throws InternalErrorException {
+		try {
+			Attribute a = getPerunBl().getAttributesManagerBl().getAttribute(sess, sponsoredUser, loginAttributeName);
+			a.setValue(login);
+			getPerunBl().getAttributesManagerBl().setAttribute(sess,sponsoredUser,a);
+		} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+			log.error("attribute {} not defined",loginAttributeName);
+		} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
+			log.error("cannot set attribute {} for sponsored user {}",loginAttributeName,sponsoredUser.getId());
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java]: Consider raising an exception, maybe InternalErrorException. That could cause whole call to abort and rollback. Without that user will think, that everything went well, which is not true.

The same problem is on more places within this pull request where you work with attributes.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java [hunk 4]
```diff
@@ -2099,6 +2085,120 @@ public void sendPasswordResetLinkEmail(PerunSession sess, Member member, String
...
+		}
+	}
+
+	@Override
+	public Member sponsorMember(PerunSession session, Vo vo, User sponsored, User sponsor) throws MemberNotExistsException, InternalErrorException, AlreadyMemberException {
+		Member sponsoredMember;
+		try {
+			sponsoredMember = getMemberByUser(session, vo, sponsored);
+			if(!sponsoredMember.isSponsored()) {
+				throw new IllegalArgumentException("member "+sponsoredMember.getId()+" is not marked as sponsored");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java]: Non of our methods (not even GUI) take into account unchecked exceptions. Can you create a checked one?

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/UsersManagerBlImpl.java [hunk 5]
```diff
@@ -1277,7 +1278,25 @@ public void validatePasswordAndSetExtSources(PerunSession sess, User user, Strin
 				} catch(UserExtSourceExistsException ex) {
 					//this is OK
 				}
-
+			} else if (loginNamespace.equals("dummy")) {
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/UsersManagerBlImpl.java]: Huh? For testing purposes I suppose. Could you please add a comment here at least?

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java (hunk 1 above)
Comment (martin-kuba) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java]: No, it is not. Concatenation of constant number of strings is by the compiler silently converted to using StringBuilder, so in fact both ways are identical in function, but mine is shorter. The change was suggested by IntelliJ IDEA.

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java (hunk 2 above)
Comment (martin-kuba) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java]: No, this indentation is correct, the first level is indented by tab, the second level matches the start of first argument, so it must use spaces. It was indented automatically anyway  :-)

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java (hunk 1 above)
Comment (licehammer) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/Auditable.java]: There's always something to learn. Thanks!

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java (hunk 2 above)
Comment (licehammer) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/User.java]: You are right. I didn't notice that.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java (hunk 3 above)
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java]: fixed

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java (hunk 4 above)
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java]: it is a checked exception. It is cz.metacentrum.perun.core.api.exceptions.IllegalArgumentException

Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java]: Ok, I have created a new MemberNotSponsoredException

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/UsersManagerBlImpl.java (hunk 5 above)
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/UsersManagerBlImpl.java]: OK

**Pull request merged**